### PR TITLE
WIP exclude browserSync and connect-history-api-fallback

### LINKED ISSUE #94: node production mode
<Have you checked issues section which may have your similar issue?>
<Please take your time filling these required information for posting issue>

## Issue Overview
If You are running npm install --production
You don`t have all needed node module to make production build on linux agent.

If You make only npm install. You are installing all packages not needed for production. Like:
browser-sync
connect-history-api-fallback

## Issue Description
gulp.config strongly banded to  gulp.config.js
- var historyApiFallback = require('connect-history-api-fallback');

And no production logic is separated. If You are following best practice installing only node mode that are required for each environment.
## Reproducables
npm install --production
gulp build --exclude-source

## Information
|                      	|                                 	|
|---------------------	|---------------------------------	|
| **Operating System** 	|  Debian |
| **Node version**     	| 4.x/5.x/orYourNodeVersionHere   	|
| **Environment**       | Browser/Mobile/WebWorker          |

### PR DESCRIPTION
**Pull request by darxx**
I what to separate browser-sync and have problems running gulp build --exclude-source --prod.
This is work in progress. Hope together we will succeed.

___
Edit by **antonybudianto**:
**TODOS**
- [x] manual testing
- [x] rebase latest master #96 

Closes #94

### CODE REVIEW

File: gulp.config.js [hunk 1]
```diff
@@ -1,8 +1,11 @@
-var historyApiFallback = require('connect-history-api-fallback')
+if (global.env === 'dev') {
+    var historyApiFallback = require('connect-history-api-fallback');
+}
 
 module.exports = function () {
     var root = '';
     var src = root + 'src/';
+    var translations = src + 'i18n/*.json';
```
Comment (antonybudianto) [inline gulp.config.js]: this one is not related right?

File: gulp.config.js [hunk 2]
```diff
@@ -31,6 +34,7 @@ module.exports = function () {
         app: 'build/app/',
         fonts: 'build/fonts',
         assetPath: 'build/assets/',
+        translate: 'build/i18n/',
```
Comment (antonybudianto) [inline gulp.config.js]: this one too

File: gulp.config.js [hunk 3]
```diff
@@ -86,6 +95,7 @@ module.exports = function () {
 
     var config = {
         root: root,
+        translations: translations,
```
Comment (antonybudianto) [inline gulp.config.js]: this one too

File: gulpfile.js [hunk 4]
```diff
@@ -1,3 +1,7 @@
+var argv = require('yargs').argv;
+
+global.env = argv.prod ? 'prod': 'dev';
```
Comment (antonybudianto) [inline gulpfile.js]: I prefer `argv.env`

File: gulp.config.js [hunk 5]
```diff
@@ -41,34 +45,39 @@ module.exports = function () {
...
-            port: 3000,
-            server: {
-                baseDir: './src/',
-                middleware: [historyApiFallback()],
-                routes: {
-                    "/node_modules": "node_modules",
-                    "/src": "src"
-                }
+
+    if (global.env === 'dev')
```
Comment (antonybudianto) [inline gulp.config.js]: Please store mode `'dev'` as constant, so it's not hardcoded and expose it to `config` var

Comment (darxx) [inline gulp.config.js]: Added 
global.env = 'dev';
global.env = argv.prod ? 'prod': 'dev';

To gulp.file.js

Comment (darxx) [inline gulp.config.js]: And I know that build is failing. I can`t find why.

Comment (darxx) [inline gulp.config.js]: C:\dev\angular2-starter>gulp build --exclude-source --prod
[16:44:39] Using gulpfile C:\dev\angular2-starter\gulpfile.js
[16:44:39] Starting 'build'...
[16:44:39] Starting 'clean-report'...
[16:44:39] Starting 'clean-ts'...
[16:44:39] Finished 'clean-report' after 6.99 ms
[16:44:39] Finished 'clean-ts' after 24 ms
[16:44:39] Starting 'tsc'...
[16:44:43] Finished 'tsc' after 4.74 s
[16:44:43] Starting 'unit-test'...
07 06 2016 16:44:49.221:INFO [karma]: Karma v0.13.22 server started at http://localhost:9876/
07 06 2016 16:44:49.229:INFO [launcher]: Starting browser PhantomJS
07 06 2016 16:44:50.816:INFO [PhantomJS 2.1.1 (Windows 8 0.0.0)]: Connected on socket /#T-pSbDyh595rPdjEAAAA with id 68686206
07 06 2016 16:44:53.273:WARN [web-server]: 404: /base/node_modules/karma-ie-launcher/test/launcher.spec
PhantomJS 2.1.1 (Windows 8 0.0.0) ERROR
  Error: XHR error (404 Not Found) loading http://localhost:9876/node_modules/karma-ie-launcher/test/launcher.spec
        Error loading http://localhost:9876/node_modules/karma-ie-launcher/test/launcher.spec


07 06 2016 16:44:53.337:WARN [web-server]: 404: /base/node_modules/karma-phantomjs-launcher/test/add.spec

**Comment by darxx** [comment]
> Please store mode 'dev' as constant, so it's not hardcoded and expose it to config var
I need predefined hard coded environment.

File: gulp.config.js (hunk 1 above)
Comment (darxx) [inline gulp.config.js]: Yes is not related. This is copy paste from working example that has localization added. If this problem related to node module separation will be solved. I can make another pull request and add localization and related unit tests.

**Comment by antonybudianto** [comment]
please update travis build script to
```
gulp build --exclude-source --env=prod
```

File: gulpfile.js [hunk 6]
```diff
@@ -1,3 +1,8 @@
+var argv = require('yargs').argv;
+
+global.env = 'dev';
```
Comment (antonybudianto) [inline gulpfile.js]: not needed

File: gulpfile.js [hunk 7]
```diff
@@ -1,3 +1,8 @@
...
+global.env = argv.prod ? 'prod': 'dev';
```
Comment (antonybudianto) [inline gulpfile.js]: `argv.prod` to `argv.env`

File: gulp.config.js (hunk 5 above)
Comment (antonybudianto) [inline gulp.config.js]: I'm tracking the error right now

**Comment by antonybudianto** [comment]
@darxx , can you edit `karma.conf.js` and go to excludedDependencies part,
```js
var excludedDependencies = [
    'systemjs', 'zone.js', 'font-awesome'
  ];
```
and add karma-ie-launcher, karma-phantomjs-launcher...

**Comment by darxx** [comment]
Hope changes will be useful to everyone. Ok I will wait. After I will fork this project again.  I`m not  spending much time on github. Don`t know how to make rebase on forked project.

**Comment by antonybudianto** [comment]
you can also pull latest master and merge it to your branch

File: .idea/jsLibraryMappings.xml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="JavaScriptLibraryMappings">
+    <file url="file://$PROJECT_DIR$" libraries="{angular2-starter node_modules}" />
+  </component>
+</project>
```
Comment (antonybudianto) [inline .idea/jsLibraryMappings.xml]: remove this file

File: .idea/libraries/angular2_starter_node_modules.xml [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+      <sourceFilesUrls>
+        <item url="file://$PROJECT_DIR$/node_modules" />
+      </sourceFilesUrls>
+    </properties>
+    <CLASSES>
+      <root url="file://$PROJECT_DIR$/node_modules" />
+    </CLASSES>
+    <SOURCES />
+  </library>
+</component>
```
Comment (antonybudianto) [inline .idea/libraries/angular2_starter_node_modules.xml]: this one too

File: gulpfile.js [hunk 10]
```diff
@@ -1,7 +1,10 @@
-var gulp = require('gulp');
-var requireDir = require('require-dir');
-var tasks = requireDir('./tasks');
+var argv = require('yargs').argv,
+    gulp = require('gulp'),
+    requireDir = require('require-dir'),
+    tasks = requireDir('./tasks');
 
+global.env = argv.env || 'dev';
+console.log(global.env);
```
Comment (antonybudianto) [inline gulpfile.js]: remove this

File: gulp.config.js [hunk 11]
```diff
@@ -1,4 +1,10 @@
-var historyApiFallback = require('connect-history-api-fallback')
+var argv = require('yargs').argv;
+
+var environment = argv.env || 'dev';
+console.log(environment);
```
Comment (antonybudianto7) [inline gulp.config.js]: remove this

File: package.json [hunk 12]
```diff
@@ -48,13 +48,8 @@
...
     "systemjs": "^0.19.29",
-    "zone.js": "^0.6.12"
-  },
-  "devDependencies": {
-    "browser-sync": "^2.12.8",
-    "codelyzer": "0.0.19",
-    "connect-history-api-fallback": "^1.2.0",
-    "del": "~2.2.0",
+    "zone.js": "^0.6.12",
+	 "del": "~2.2.0",
```
Comment (antonybudianto7) [inline package.json]: sorry about being picky :smile: , please remove the tab/additional spaces here

**Comment by antonybudianto7** [comment]
the rest LGTM, will check it out tonight

**Comment by antonybudianto** [comment]
Thanks for the PR! @darxx
